File: src/test/java/com/wezaam/withdrawal/service/WithdrawalServiceTest.java
package com.wezaam.withdrawal.service;

import com.wezaam.withdrawal.exception.ForbiddenException;
import com.wezaam.withdrawal.exception.NotFoundException;
import com.wezaam.withdrawal.exception.TransactionException;
import com.wezaam.withdrawal.model.PaymentMethod;
import com.wezaam.withdrawal.model.Withdrawal;
import com.wezaam.withdrawal.model.enums.WithdrawalStatus;
import com.wezaam.withdrawal.repository.WithdrawalRepository;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Optional;
import java.util.concurrent.ExecutorService;

import static com.wezaam.withdrawal.service.MocksHelper.mockPaymentMethod;
import static com.wezaam.withdrawal.service.MocksHelper.mockUser;
import static java.util.Arrays.asList;
import static org.junit.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.*;


@RunWith(MockitoJUnitRunner.class)
public class WithdrawalServiceTest {

    @Mock
    private WithdrawalRepository withdrawalRepository;
    @Mock
    private WithdrawalProcessingService withdrawalProcessingService;
    @Mock
    private PaymentMethodService paymentMethodService;
    @Mock
    private EventsService eventsService;
    @Mock
    private UserService userService;
    @Mock
    private ExecutorService executorService;
    @InjectMocks
    private WithdrawalService withdrawalService;

    @Test(expected = NotFoundException.class)
    public void testCreateWithdrawalWithANoExistingUser() {
        Long userId = 1L;
        when(userService.findById(userId)).thenThrow(new NotFoundException(""));
        withdrawalService.create(userId, 2L, 100D, "ASAP");
    }

    @Test(expected = ForbiddenException.class)
    public void testCreateWithdrawalWithAnAmountHigherThanMaxAmount() {
        Long userId = 1L;
        Long paymentMethodId = 2L;
        when(userService.findById(userId)).thenReturn(mockUser(100d, paymentMethodId));
        withdrawalService.create(userId, paymentMethodId, 1000D, "ASAP");
    }

    @Test(expected = NotFoundException.class)
    public void testCreateWithdrawalWithANoExistingPaymentMethod() {
        Long userId = 1L;
        Long paymentMethodId = 2L;
        when(userService.findById(userId)).thenReturn(mockUser(100d, paymentMethodId));
        when(paymentMethodService.findById(paymentMethodId)).thenThrow(new NotFoundException(""));
        withdrawalService.create(userId, paymentMethodId, 100D, "ASAP");
    }

    @Test(expected = ForbiddenException.class)
    public void testCreateWithdrawalWithADifferentPaymentMethod() {
        Long userId = 1L;
        Long userPaymentMethodId = 2L;
        Long withdrawalPaymentMethodId = 3L;
        Double withdrawalAmount = 100D;
        Double maxWithdrawalAmount = 100D;

        when(userService.findById(userId)).thenReturn(mockUser(maxWithdrawalAmount, userPaymentMethodId));
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod(withdrawalPaymentMethodId));

        withdrawalService.create(userId, withdrawalPaymentMethodId, withdrawalAmount, "ASAP");
    }

    @Test
    public void testCreateWithdrawalStartingAsap() {
        Long userId = 1L;
        Long userPaymentMethodId = 2L;
        Long withdrawalPaymentMethodId = 2L;
        Double withdrawalAmount = 100D;
        Double maxWithdrawalAmount = 100D;
        Long withdrawalId = 1L;

        when(userService.findById(userId)).thenReturn(mockUser(maxWithdrawalAmount, userPaymentMethodId));
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod(withdrawalPaymentMethodId));
        when(withdrawalRepository.findById(any())).thenReturn(Optional.of(mockWithdrawal(withdrawalId, WithdrawalStatus.PROCESSING)));

        Withdrawal result = withdrawalService.create(userId, withdrawalPaymentMethodId, withdrawalAmount, "ASAP");

        assertEquals(WithdrawalStatus.PROCESSING, result.getStatus());
    }

    @Test
    public void testCreateWithdrawalStartingInTheFuture() {
        Long userId = 1L;
        Long userPaymentMethodId = 2L;
        Long withdrawalPaymentMethodId = 2L;
        Double withdrawalAmount = 100D;
        Double maxWithdrawalAmount = 100D;
        Long withdrawalId = 1L;

        when(userService.findById(userId)).thenReturn(mockUser(maxWithdrawalAmount, userPaymentMethodId));
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod(withdrawalPaymentMethodId));
        when(withdrawalRepository.findById(any())).thenReturn(Optional.of(mockWithdrawal(withdrawalId, WithdrawalStatus.PENDING)));

        Instant tomorrow = Instant.now().plus(1, ChronoUnit.DAYS);
        Withdrawal result = withdrawalService.create(userId, withdrawalPaymentMethodId, withdrawalAmount, tomorrow.toString());

        assertEquals(WithdrawalStatus.PENDING, result.getStatus());
    }

    private static Withdrawal mockWithdrawal(Long withdrawalId, WithdrawalStatus status) {
        Withdrawal withdrawal = new Withdrawal();
        withdrawal.setId(withdrawalId);
        withdrawal.setStatus(status);
        return withdrawal;
    }

    @Test
    public void runShouldCallExecutorServiceForEachWithdrawal() {
        Withdrawal mock1 = mockWithdrawal(1L, WithdrawalStatus.PENDING);
        Withdrawal mock2 = mockWithdrawal(2L, WithdrawalStatus.PENDING);
        Withdrawal mock3 = mockWithdrawal(3L, WithdrawalStatus.PENDING);
        when(withdrawalRepository.findAllByExecuteAtBeforeAndStatus(any(), eq(WithdrawalStatus.PENDING)))
                .thenReturn(asList(mock1, mock2, mock3));

        withdrawalService.run();

        verify(executorService, times(3)).submit(any(Runnable.class));
    }

    @Test
    public void testProcess() throws TransactionException {
        Long transactionId = 123L;
        Long withdrawalPaymentMethodId = 2L;
        Double withdrawalAmount = 100d;
        Withdrawal mockWithdrawal = mockWithdrawal(1L, WithdrawalStatus.PENDING);
        mockWithdrawal.setPaymentMethodId(withdrawalPaymentMethodId);
        mockWithdrawal.setAmount(withdrawalAmount);
        PaymentMethod mockPaymentMethod = mockPaymentMethod(withdrawalPaymentMethodId);
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod);
        when(withdrawalProcessingService.sendToProcessing(withdrawalAmount, mockPaymentMethod)).thenReturn(transactionId);

        withdrawalService.process(mockWithdrawal);

        Withdrawal expectedWithdrawalAfterTransaction = mockWithdrawal(1L, WithdrawalStatus.PROCESSING);
        expectedWithdrawalAfterTransaction.setPaymentMethodId(withdrawalPaymentMethodId);
        expectedWithdrawalAfterTransaction.setAmount(withdrawalAmount);
        expectedWithdrawalAfterTransaction.setTransactionId(transactionId);
        verify(withdrawalRepository, times(2)).save(expectedWithdrawalAfterTransaction);
    }

    @Test
    public void testProcessWithTransactionException() throws TransactionException {
        Long withdrawalPaymentMethodId = 2L;
        Double withdrawalAmount = 100d;
        Withdrawal mockWithdrawal = mockWithdrawal(1L, WithdrawalStatus.PENDING);
        mockWithdrawal.setPaymentMethodId(withdrawalPaymentMethodId);
        mockWithdrawal.setAmount(withdrawalAmount);
        PaymentMethod mockPaymentMethod = mockPaymentMethod(withdrawalPaymentMethodId);
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod);
        when(withdrawalProcessingService.sendToProcessing(withdrawalAmount, mockPaymentMethod)).thenThrow(new TransactionException());

        withdrawalService.process(mockWithdrawal);

        Withdrawal expectedWithdrawalAfterTransaction = mockWithdrawal(1L, WithdrawalStatus.FAILED);
        expectedWithdrawalAfterTransaction.setPaymentMethodId(withdrawalPaymentMethodId);
        expectedWithdrawalAfterTransaction.setAmount(withdrawalAmount);
        verify(withdrawalRepository, times(2)).save(expectedWithdrawalAfterTransaction);
    }

    @Test
    public void testProcessWithException() throws TransactionException {
        Long withdrawalPaymentMethodId = 2L;
        Double withdrawalAmount = 100d;
        Withdrawal mockWithdrawal = mockWithdrawal(1L, WithdrawalStatus.PENDING);
        mockWithdrawal.setPaymentMethodId(withdrawalPaymentMethodId);
        mockWithdrawal.setAmount(withdrawalAmount);
        PaymentMethod mockPaymentMethod = mockPaymentMethod(withdrawalPaymentMethodId);
        when(paymentMethodService.findById(withdrawalPaymentMethodId)).thenReturn(mockPaymentMethod);
        when(withdrawalProcessingService.sendToProcessing(withdrawalAmount, mockPaymentMethod)).thenThrow(new RuntimeException());

        withdrawalService.process(mockWithdrawal);

        Withdrawal expectedWithdrawalAfterTransaction = mockWithdrawal(1L, WithdrawalStatus.INTERNAL_ERROR);
        expectedWithdrawalAfterTransaction.setPaymentMethodId(withdrawalPaymentMethodId);
        expectedWithdrawalAfterTransaction.setAmount(withdrawalAmount);
        verify(withdrawalRepository, times(2)).save(expectedWithdrawalAfterTransaction);
    }

    @Test
    public void testFindAll() {
        withdrawalService.findAll();
        verify(withdrawalRepository).findAll();
    }

    @Test(expected = NotFoundException.class)
    public void testFindByIdWhenWithdrawalDoesNotExists() {
        Long withdrawalId = 1l;
        when(withdrawalRepository.findById(withdrawalId)).thenThrow(new NotFoundException(""));

        withdrawalService.findById(withdrawalId);
    }

    @Test
    public void testFindByIdWhenWithdrawalExists() {
        Long withdrawalId = 1l;
        when(withdrawalRepository.findById(withdrawalId)).thenReturn(Optional.of(mockWithdrawal(withdrawalId, WithdrawalStatus.PROCESSING)));

        Withdrawal result = withdrawalService.findById(withdrawalId);
        assertEquals(withdrawalId, result.getId());
    }
}

File: src/test/java/service/WithdrawalServiceTest.kt
package service

import com.wezaam.withdrawal.dto.WithdrawalInformationDTO
import com.wezaam.withdrawal.exception.TransactionException
import com.wezaam.withdrawal.model.PaymentMethod
import com.wezaam.withdrawal.model.WithdrawType
import com.wezaam.withdrawal.model.WithdrawalEntity
import com.wezaam.withdrawal.model.WithdrawalStatus
import com.wezaam.withdrawal.repository.WithdrawalRepository
import com.wezaam.withdrawal.repository.WithdrawalScheduledRepository
import com.wezaam.withdrawal.service.EventsService
import com.wezaam.withdrawal.service.WithdrawalFactory
import com.wezaam.withdrawal.service.WithdrawalProcessingService
import com.wezaam.withdrawal.service.WithdrawalService
import org.junit.Before
import org.junit.Test

import org.junit.Assert.*
import org.mockito.Mockito
import org.mockito.Mockito.times
import java.time.Instant

class WithdrawalServiceTest {

    private val withdrawalRepository = Mockito.mock(WithdrawalRepository::class.java)
    private val withdrawalScheduleRepository = Mockito.mock(WithdrawalScheduledRepository::class.java)
    private val withdrawalFactory = WithdrawalFactory(withdrawalRepository, withdrawalScheduleRepository)
    private val notifierService = Mockito.mock(EventsService::class.java)
    private val processService = WithdrawalProcessingService()
    private val withdrawalService = WithdrawalService (withdrawalFactory, processService, notifierService)
    private val withdrawalInformationDTO = WithdrawalInformationDTO(
            23,
            0,
            500.0,
            Instant.now(),
            Instant.now(),
            3,
            PaymentMethod(1,"test account"),
            WithdrawalStatus.PENDING,
            WithdrawType.ASAP)
    private val withdrawalEntity = WithdrawalEntity(
            withdrawalInformationDTO.id,
            withdrawalInformationDTO.transactionId,
            withdrawalInformationDTO.amount,
            Instant.now(),
            withdrawalInformationDTO.userId,
            withdrawalInformationDTO.paymentMethod,
            WithdrawalStatus.PENDING)
    private val updatedEntity = WithdrawalEntity(
            System.nanoTime(),
            withdrawalInformationDTO.transactionId,
            withdrawalInformationDTO.amount,
            Instant.now(),
            withdrawalInformationDTO.userId,
            withdrawalInformationDTO.paymentMethod,
            WithdrawalStatus.PROCESSING)

    @Before
    fun setUp() {
    }

    @Test
    fun create() {
        Mockito.`when`(withdrawalRepository.save(any(WithdrawalEntity::class.java))).thenReturn(withdrawalEntity)
        val result = withdrawalService.create(withdrawalInformationDTO)
        assertNotNull(result)
        Mockito.verify(withdrawalRepository, times(1)).save(any(WithdrawalEntity::class.java))
    }

    @Test
    fun process() {
        Mockito.`when`(withdrawalRepository.save(any(WithdrawalEntity::class.java))).thenReturn(updatedEntity)
        val result = withdrawalService.process(withdrawalInformationDTO)
        assertNotNull(result)
        assertNotNull(result.transactionId)
        assertTrue(result.status == WithdrawalStatus.PROCESSING)
        Mockito.verify(withdrawalRepository, times(1)).save(any(WithdrawalEntity::class.java))
        Mockito.verify(notifierService, times(1)).send(any(WithdrawalInformationDTO::class.java))
    }

    @Test(expected = TransactionException::class)
    fun testThatProcessThrowsException(){
        Mockito.`when`(withdrawalRepository.save(any(WithdrawalEntity::class.java)))
                .thenThrow(TransactionException("Test message"))
        val result = withdrawalService.process(withdrawalInformationDTO)
        assertNotNull(result)
        assertNotNull(result.transactionId)
        assertTrue(result.status == WithdrawalStatus.FAILED)
    }

    private fun <T> any(type : Class<T>): T {
        Mockito.any(type)
        return uninitialized()
    }

    private fun <T> uninitialized(): T = null as T
}